### PR TITLE
support MinMax hyperrectangle for MergeTreeIndexSet

### PR DESCRIPTION
**Pull request by AntiTopQuark**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support minmax hyperrectangle for Set indices

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit e29c8f9aeefb188efc2052d08f123e45b5b3309b with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65676/e29c8f9aeefb188efc2052d08f123e45b5b3309b/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Cloud fork sync (only for ClickHouse Inc. employees)</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by AntiTopQuark** [comment]
Sorry.
I marked the PR as a draft until I fix failed CI tests.

**Comment by AntiTopQuark** [comment]
Hi, @nickitat . It seems the test case is unstable. Should I retry it?

**Comment by nickitat** [comment]
> It seems the test case is unstable. Should I retry it?

which one exactly do you mean?

**Comment by AntiTopQuark** [comment]
> > It seems the test case is unstable. Should I retry it?
> 
> which one exactly do you mean?

Performance Comparison: 
https://s3.amazonaws.com/clickhouse-test-reports/65676/aa1ee412f17ff15b4e14ef936fbfd408c2f4af78/performance_comparison__release__[4_4]/report.html

File: src/Storages/MergeTree/MergeTreeIndexSet.cpp [hunk 1]
```diff
@@ -85,6 +87,15 @@ void MergeTreeIndexGranuleSet::serializeBinary(WriteBuffer & ostr) const
         serialization->serializeBinaryBulkWithMultipleStreams(column, 0, size(), settings, state);
         serialization->serializeBinaryBulkStateSuffix(settings, state);
     }
+
+    for (size_t i = 0; i < num_columns; ++i)
```
Comment (nickitat) [inline src/Storages/MergeTree/MergeTreeIndexSet.cpp:91]: let's add tests to ensure backward and forward compatibility. newer version should be able to read old parts, old versions should be able to read new (just inserted by the new version) parts.
for the reference you may use the tests under `tests/integration/test_backward_compatibility`

File: src/Storages/MergeTree/MergeTreeIndexSet.cpp [hunk 2]
```diff
@@ -546,7 +605,7 @@ MergeTreeIndexAggregatorPtr MergeTreeIndexSet::createIndexAggregator(const Merge
 MergeTreeIndexConditionPtr MergeTreeIndexSet::createIndexCondition(
     const ActionsDAGPtr & filter_actions_dag, ContextPtr context) const
 {
-    return std::make_shared<MergeTreeIndexConditionSet>(index.name, index.sample_block, max_rows, filter_actions_dag, context);
+    return std::make_shared<MergeTreeIndexConditionSet>(index.name, index.sample_block, max_rows, filter_actions_dag, context, index);
```
[the remark this answers is not in the document]
Comment (AntiTopQuark) [reply src/Storages/MergeTree/MergeTreeIndexSet.cpp:608]: done, thank you.

File: src/Storages/MergeTree/MergeTreeIndexSet.cpp (hunk 1 above)
Comment (AntiTopQuark) [reply src/Storages/MergeTree/MergeTreeIndexSet.cpp:91]: The current implementation does indeed have compatibility issues. 
the maximum and minimum values are recalculated after deserializing the columns, and then they are added as a Range to `set_hyperrectangle`. `set_hyperrectangle` does not participate in serialization and deserializationThis approach avoids serialization compatibility issues.

File: src/Storages/MergeTree/MergeTreeIndexSet.cpp [hunk 3]
```diff
@@ -117,6 +128,25 @@ void MergeTreeIndexGranuleSet::deserializeBinary(ReadBuffer & istr, MergeTreeInd
...
+    {
+        const DataTypePtr & type = block.getByPosition(i).type;
+        auto serialization = type->getDefaultSerialization();
+
+        serialization->deserializeBinary(min_val, istr, {});
+        serialization->deserializeBinary(max_val, istr, {});
+
+        // NULL_LAST
+        if (min_val.isNull())
+            min_val = POSITIVE_INFINITY;
```
[the remark this answers is not in the document]
Comment (AntiTopQuark) [reply src/Storages/MergeTree/MergeTreeIndexSet.cpp:145]: This refers to the implementation of the MinMax index. In the new version, set_hyperrectangle is not involved in serialization and deserialization, and the corresponding code has been removed.

File: src/Storages/MergeTree/MergeTreeIndexSet.cpp [hunk 4]
```diff
@@ -117,6 +128,25 @@ void MergeTreeIndexGranuleSet::deserializeBinary(ReadBuffer & istr, MergeTreeInd
...
+        if (max_val.isNull())
+            max_val = POSITIVE_INFINITY;
+        set_hyperrectangle.emplace_back(min_val, true, max_val, true);
```
[the remark this answers is not in the document]
Comment (AntiTopQuark) [reply src/Storages/MergeTree/MergeTreeIndexSet.cpp:148]: This refers to the implementation of the MinMax index. In the new version, set_hyperrectangle is not involved in serialization and deserialization, and the corresponding code has been removed.

**Review by nickitat — APPROVED** [review]
lgtm

**Comment by AntiTopQuark** [comment]
@nickitat @Algunenano It looks like this PR can be merged now. Is there anything else I need to do?

**Pull request merged**